### ISSUE TITLE
Bug: $this->request->getUri()->getPath() is not compatible with PSR-7

### ISSUE THREAD

**Issue by kenjis** [codeigniter4/CodeIgniter4#7123]
CodeIgniter4 Version: v4.3.1

`$this->request->getUri()->getPath()` returns the route (URI path relative to baseURL).

```
app.baseURL = 'http://localhost:8888/ci431/public/'
```
When navigating to `http://localhost:8888/ci431/public/test?a=b`,

```php
$uri = $this->request->getUri();
echo (string) $uri;   // "http://localhost:8888/ci431/public/test?a=b" → Okay
echo $uri->getPath(); // "test" → NG. It should be "/ci431/public/test" when following PSR-7
```

Ref https://www.php-fig.org/psr/psr-7/

Related #5930

**Comment by kenjis** [comment]
>  I will push again for a separate class to handle internal URIs; I think we need it despite the complexity it adds.
https://github.com/codeigniter4/CodeIgniter4/issues/7124#issuecomment-1383940519

**Comment by kenjis** [comment]
Navigate to `http://localhost:8888/ci431/public/test?a=b`:
```
    uri_string(): test
uri_string(true): test

               current_url(): http://localhost:8888/ci431/public/index.php/test
current_url(true)->getPath(): /ci431/public/index.php/test
  (string) current_url(true): http://localhost:8888/ci431/public/index.php/test?a=b

  (string) $this->request->getUri(): http://localhost:8888/ci431/public/test?a=b
$this->request->getUri()->getPath(): test
```

**Comment by kenjis** [comment]
Navigate to `http://localhost:8888/ci431/public/index.php/test?a=b`:
```
    uri_string(): test
uri_string(true): test

               current_url(): http://localhost:8888/ci431/public/index.php/test
current_url(true)->getPath(): /ci431/public/index.php/test
  (string) current_url(true): http://localhost:8888/ci431/public/index.php/test?a=b

  (string) $this->request->getUri(): http://localhost:8888/ci431/public/test?a=b
$this->request->getUri()->getPath(): test
```

**Comment by kenjis** [comment]
`current_url()` returns the URL with `index.php`.
But `$this->request->getUri()` does not have `index.php`.

I think when `current_url()` has `index.php`, the URI object also should have `index.php`.

**Comment by kenjis** [comment]
From https://github.com/codeigniter4/CodeIgniter4/pull/7233#issuecomment-1426700790

The inconsistency of URI Segments between `$this->request->getUri()` and `current_url(true)`.

```php
<?php

namespace App\Controllers;

class Home extends BaseController
{
    public function index()
    {
        var_dump($this->request->getUri()->getSegments());
        var_dump($this->request->getUri()->getSegment(1));
        $uri = current_url(true);
        var_dump($uri->getSegments());
        var_dump($uri->getSegment(1));
    }
}
```
```
/Applications/MAMP/htdocs/ci431/app/Controllers/Home.php:9:
array (size=3)
  0 => string 'one' (length=3)
  1 => string 'two' (length=3)
  2 => string 'three' (length=5)

/Applications/MAMP/htdocs/ci431/app/Controllers/Home.php:10:string 'one' (length=3)

/Applications/MAMP/htdocs/ci431/app/Controllers/Home.php:12:
array (size=6)
  0 => string 'ci431' (length=5)
  1 => string 'public' (length=6)
  2 => string 'index.php' (length=9)
  3 => string 'one' (length=3)
  4 => string 'two' (length=3)
  5 => string 'three' (length=5)

/Applications/MAMP/htdocs/ci431/app/Controllers/Home.php:13:string 'ci431' (length=5)
```

**Comment by kenjis** [comment]
About PSR-7 URI::getPath(): https://github.com/codeigniter4/CodeIgniter4/pull/7252#issuecomment-1434060635

**Comment by kenjis** [comment]
Navigate to `http://localhost:8888/ci431/public/`:
```
    uri_string(): /
uri_string(true): 

               current_url(): http://localhost:8888/ci431/public/index.php/
current_url(true)->getPath(): /ci431/public/index.php/
  (string) current_url(true): http://localhost:8888/ci431/public/index.php/

  (string) $this->request->getUri(): http://localhost:8888/ci431/public/
$this->request->getUri()->getPath(): /
```

**Comment by kenjis** [comment]
Navigate to `http://localhost:8888/ci431/public/index.php`:
```
    uri_string(): /
uri_string(true): 

               current_url(): http://localhost:8888/ci431/public/index.php/
current_url(true)->getPath(): /ci431/public/index.php/
  (string) current_url(true): http://localhost:8888/ci431/public/index.php/

  (string) $this->request->getUri(): http://localhost:8888/ci431/public/
$this->request->getUri()->getPath(): /
```

**Comment by kenjis** [comment]
> Finally, I think URI and SiteURL can be made fully compatible (and having some extended functions) with PSR-7.
https://github.com/codeigniter4/CodeIgniter4/pull/7252#issuecomment-1430566404

See https://github.com/codeigniter4/CodeIgniter4/pull/7252#issuecomment-1434202058
It seems if we make our URI PSR-7 compatible, we need to distinguish the path `/foo/bar` and `foo/bar`.
When the baseURL is `http://localhost/ci4/`, `$uri->withPath('/foo/bar')` will be `http://localhost/foo/bar`.
`$uri->withPath('foo/bar')` will be `http://localhost/ci4/foo/bar`.

But we do not distinguish the route path `/foo/bar` and `foo/bar`.
They are exactly the same.

**Comment by kenjis** [comment]
Closed by #7282
